### PR TITLE
soc: xtensa: imx8: Split generic i.MX8 into i.MX8QM and i.MX8QXP

### PR DESCRIPTION
**Pull request by LaurentiuM1234**
As things stand now, the `i.MX8` SoC doesn't have any NXP HAL files to back it up. As such, the Zephyr native drivers cannot be used.

The `i.MX8` is used to represent `i.MX8QM` and `i.MX8QXP`, which have different NXP HAL files. As such, splitting `i.MX8` into its 2 variants is necessary.

Apart from this, this patch series does some cleanups and adds support for operations such as clock management, pinctrl management and serial communication.

### CODE REVIEW

File: boards/xtensa/nxp_adsp_imx8x/nxp_adsp_imx8x-pinctrl.dtsi
Comment (gmarull) [inline boards/xtensa/nxp_adsp_imx8x/nxp_adsp_imx8x-pinctrl.dtsi:15]: omit-if-no-ref is meant for autogenerated files, why is this defined here?

File: boards/xtensa/nxp_adsp_imx8/nxp_adsp_imx8_defconfig
Comment (gmarull) [inline boards/xtensa/nxp_adsp_imx8/nxp_adsp_imx8_defconfig:19]: make sure drivers using pinctrl `select PINCTRL` instead, see e.g. https://github.com/zephyrproject-rtos/zephyr/pull/62815

File: dts/bindings/pinctrl/nxp,imx-iomuxc-scu.yaml
Comment (gmarull) [inline dts/bindings/pinctrl/nxp,imx-iomuxc-scu.yaml:29]: why are some properties being encoded in non-standard way? ie pulls, strength, etc.

**Review by gmarull — DISMISSED** [review]

File: dts/bindings/pinctrl/nxp,imx-iomuxc-scu.yaml
Comment (LaurentiuM1234) [reply dts/bindings/pinctrl/nxp,imx-iomuxc-scu.yaml:29]: The standard way of encoding these properties would work really well with drive strength and pull selection but not so much with the I/O configuration (`sw_config`) and the low power configuration (`lp_config`). For instance, to support all I/O configuration cases you'd have to use a mix of the `drive-open-drain`, `input-enable`, and `output-enable` properties (e.g: no property => NORMAL mode, `drive-open-drain` => OD mode with input or output buffer enabled, `drive-open-drain` + `input-enable` => OD mode with input buffer enabled (output buffer may be enabled or not) etc...). For me  this seems much messier to do rather than just using the macros provided in the pinctrl header.
Also, in the case of `lp_config`, there seems to be only one configuration supported? (i.e: `bias-bus-hold`)

I guess the "better" alternative would be to just encode all of the configurations in a single field in the pinmux property (as Linux seems to do) but if this is not a deal-breaker it can be addressed later on I guess.

My knowledge of the pad configuration is rather modest so please let me know if I've got something wrong.

File: boards/xtensa/nxp_adsp_imx8x/nxp_adsp_imx8x-pinctrl.dtsi
Comment (LaurentiuM1234) [reply boards/xtensa/nxp_adsp_imx8x/nxp_adsp_imx8x-pinctrl.dtsi]: Not sure why I decided to keep the `/omit-if-no-ref/` statements as they're not necessary at this point. Either way, I've removed them.

File: dts/bindings/pinctrl/nxp,imx-iomuxc-scu.yaml
Comment (gmarull) [reply dts/bindings/pinctrl/nxp,imx-iomuxc-scu.yaml:29]: This is a requirement for new bindings, see https://docs.zephyrproject.org/latest/hardware/pinctrl/index.html#pin-control-drivers

Exceptions include implementations following an alternative convention followed elsewhere (eg Linux), provided there's a 1:1 matching (no Zephyr tweaks).

**Review by iuliana-prodan — DISMISSED** [review]

File: dts/bindings/pinctrl/nxp,imx-iomuxc-scu.yaml
Comment (LaurentiuM1234) [reply dts/bindings/pinctrl/nxp,imx-iomuxc-scu.yaml]: Thanks for letting me know. Should be better now.

**Review by iuliana-prodan — DISMISSED** [review]

File: dts/bindings/pinctrl/nxp,imx-iomuxc-scu.yaml
Comment (danieldegrasse) [reply dts/bindings/pinctrl/nxp,imx-iomuxc-scu.yaml]: Looking at the RM, it seems that the underlying IP behind the SCU is the IOMUXD- couldn't we declare custom properties where needed for certain pin configuration options? This is what was done for the IOMUXC peripheral: https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/pinctrl/nxp%2Cmcux-rt-pinctrl.yaml. Please let me know if I'm missing something here, I'm not very familiar with this SOC or the SCU.

Comment (LaurentiuM1234) [reply dts/bindings/pinctrl/nxp,imx-iomuxc-scu.yaml]: I think this would also work, thanks for the suggestion! Currently, I've dropped the feature of allowing pad configuration as it was a bit of a headache and we weren't really using non-default configurations. This is more of a personal taste I guess but if possible I'd rather encode all of the configurations in a 32-bit field passed through the `pinmux` property instead of having to specify all of the standard properties (and custom ones) just like we do in Linux (seems more simple this way I guess?) (but this can change once I see the actual implementation). Anyways, at the moment this feature is more of a `nice to have` feature instead of a `must have` one so we'll deal with this later if everyone finds this acceptable.

File: west.yml
Comment (TangleZ) [inline west.yml:196]: Is here should be a commit num?

Comment (iuliana-prodan) [reply west.yml:196]: Yes, it should be, but for testing we can add a pull request until that PR is merged.

This PR has a DNM (do not merge) label and it cannot be merged until this label is removed (this is don't automatically). This label is added, also automatically, because we have a PR in west.yml, instead of a commit id.

image

**Comment by dbaluta** [comment]
@LaurentiuM1234  can you let me know the PR that this PR depends on? Also, we should rebase this PR on HEAD as it has some conflicts.

**Comment by LaurentiuM1234** [comment]
> can you let me know the PR that this PR depends on?

https://github.com/zephyrproject-rtos/hal_nxp/pull/290

> Also, we should rebase this PR on HEAD as it has some conflicts.

Should be fixed now.

**Comment by dbaluta** [comment]
@gmarull all comments addressed.

**Comment by iuliana-prodan** [comment]
@LaurentiuM1234  can you please rebase https://github.com/zephyrproject-rtos/hal_nxp/pull/290 on latest nxp_hal?
This will fix the CI which is failing on building `openamp_rsc_table` sample for `nxp_adsp_imx8m`.
From your PR #290 is missing this commit https://github.com/zephyrproject-rtos/hal_nxp/commit/361ccc5962dfc691f6134d73605fed063c4d26cd which is needed for the above sample.

**Comment by LaurentiuM1234** [comment]
> @LaurentiuM1234 can you please rebase [zephyrproject-rtos/hal_nxp#290](https://github.com/zephyrproject-rtos/hal_nxp/pull/290) on latest nxp_hal? This will fix the CI which is failing on building `openamp_rsc_table` sample for `nxp_adsp_imx8m`. From your PR 290 is missing this commit [zephyrproject-rtos/hal_nxp@361ccc5](https://github.com/zephyrproject-rtos/hal_nxp/commit/361ccc5962dfc691f6134d73605fed063c4d26cd) which is needed for the above sample.

I propose waiting until the NXP HAL PR is merged since we won't be able to merge this one until then anyways and any approves I get will most likely get dismissed once I update `west.yaml` to point to a commit hash instead of the PR. This way we can also avoid running the test suite twice.

**Comment by mmahadevan108** [comment]
@LaurentiuM1234 . HAL PR has been merged. Kindly update the `west.yml` file

**Comment by LaurentiuM1234** [comment]
> @LaurentiuM1234 . HAL PR has been merged. Kindly update the `west.yml` file

Done.

All dependencies have been merged and all tests are passing. @iuliana-prodan could you please review again? Thanks!

**Review by dbaluta — DISMISSED** [review]

**Review by iuliana-prodan — DISMISSED** [review]

File: dts/xtensa/nxp/nxp_imx8.dtsi
Comment (LaurentiuM1234) [reply dts/xtensa/nxp/nxp_imx8.dtsi:29]: The dummy INTC is going to be required anyways once we enable the IRQSTEER driver and the multi-level interrupts so in the end having LPUART's `interrupts` property as optional wouldn't change much. Either way, if this becomes a problem it should be addressed in another PR.

**Note**: the dummy INTID passed through LPUART's `interrupts` property is just a bad way of avoiding a dependency on https://github.com/zephyrproject-rtos/zephyr/pull/62776 and will be solved once said PR is merged.

**Review by iuliana-prodan — APPROVED** [review]

**Review by mmahadevan108 — APPROVED** [review]

**Pull request merged**
